course5/homework04/homework04.cc

```diff
@@ -0,0 +1,115 @@
+// Example program
+#include <string>
+#include <cstdio>
+#include <cstdlib>
+
+class Endl
+{
+};
+
+
+class Stream
+{    
+private:
+        std::FILE * file;
+public:
+        static Endl endl;
+
+        Stream() = delete;
+
+        Stream(FILE * file) : file(file) { }
+
+        Stream(const char * filename)
+	{
+                file = std::fopen(filename, "w");
+	}
+
+        Stream(Stream&& source)
+        {
+                file = source.file;
+                source.file = nullptr;
+        }
+
+        Stream(const Stream& other) = delete;
+
+	~Stream()
+        {
+                if (file && file != stdout)
+                    std::fclose(file);
+	}
+
+        Stream& operator = (Stream&& stream)
+        {
+                file = stream.file;
+                stream.file = nullptr;
+                return *this;
+        }
+
+	Stream& operator << (std::string value)
+	{
+		std::fprintf(file, "%s", value.c_str());
+                return *this;
+	}
+
+        Stream& operator << (bool value)
+        {
+                std::fprintf(file, "%s", value ? "true" : "false");
+                return *this;
+        }
+
+        Stream& operator << (int value)
+        {
+                std::fprintf(file, "%d", value);
+                return *this;
+        }
+
+        Stream& operator << (float value)
+        {
+                //std::string format = std::string("%.") + std::to_string(precision).c_str() + std::string("f");
+                std::fprintf(file, "%f", value);
+                return *this;
+        }
+
+        Stream& operator << (double value)
+        {
+                std::fprintf(file, "%f", value);
+                return *this;
+        }
+
+        Stream& operator << (Endl)
+        {
+                std::fprintf(file, "\n");
+                std::fflush(file);
+                return *this;
+        }
+
+};
+
+void test(Stream& cout)
+{
+        // comment indicates the expected input
+        cout << 7 << 5 << Stream::endl; // 75
+        cout << true;// true
+        cout << 1.0/3.0;// 0.33
+
+        std::string s = "world!";
+        cout << "hello, "; // hello,
+        cout << s; // world!
+        cout << Stream::endl;
+
+        float f = 6;
+        cout << f; // 6
+        cout << Stream::endl;
```

Comment:
while implementing this test I found a bug. I haven't implemented print method taking char * as a parameter. apparently compiler casted char * to bool and the output was incorrect (true instead of hello, ). Now I wonder why was compiler choosing bool, not std::string overload?